Search.setIndex({"docnames": ["landing-page", "notebooks/ancestry_informative_markers", "notebooks/configuring_the_workflow", "notebooks/contributing", "notebooks/differential_expression", "notebooks/gene_families", "notebooks/genetic_diversity", "notebooks/input_data", "notebooks/installation", "notebooks/karyotyping", "notebooks/quality_control", "notebooks/rna-seq-pop", "notebooks/running_the_workflow", "notebooks/selection", "notebooks/troubleshooting", "notebooks/variant_calling"], "filenames": ["landing-page.md", "notebooks/ancestry_informative_markers.ipynb", "notebooks/configuring_the_workflow.ipynb", "notebooks/contributing.ipynb", "notebooks/differential_expression.ipynb", "notebooks/gene_families.ipynb", "notebooks/genetic_diversity.ipynb", "notebooks/input_data.ipynb", "notebooks/installation.ipynb", "notebooks/karyotyping.ipynb", "notebooks/quality_control.ipynb", "notebooks/rna-seq-pop.ipynb", "notebooks/running_the_workflow.ipynb", "notebooks/selection.ipynb", "notebooks/troubleshooting.ipynb", "notebooks/variant_calling.ipynb"], "titles": ["RNA-Seq-Pop", "Ancestry", "Configuring the workflow", "Contributing to the workflow", "Differential expression", "Gene family heatmaps", "Genetic diversity", "Input files", "Installation", "Karyotyping", "Quality control", "<strong>RNA-Seq-Pop modules</strong>", "Running the workflow", "Selection", "Troubleshooting", "Variant calling"], "terms": {"welcom": 0, "document": 0, "i": [0, 2, 7, 8, 12], "comput": [0, 8], "pipelin": 0, "analys": [0, 2, 7, 8], "illumina": 0, "sequenc": [0, 7], "data": [0, 2, 5, 7, 12], "ani": [0, 3, 8], "organ": [0, 2, 7], "As": [0, 8], "well": [0, 7], "perform": [0, 3, 7, 12], "standard": [0, 12], "transcriptom": [0, 2, 7], "differenti": [0, 2], "express": [0, 2], "also": [0, 2, 14], "call": [0, 2, 5, 7], "genet": [0, 2], "polymorph": 0, "extract": 0, "popul": [0, 2], "genom": [0, 2, 7], "signal": 0, "thi": [0, 2, 3, 7, 8], "jupyt": 0, "book": 0, "help": [0, 14], "you": [0, 2, 3, 8, 12, 14], "instal": [0, 12], "configur": [0, 7, 12], "appli": 0, "your": [0, 2, 3, 7, 8, 12], "own": 0, "sanjai": 0, "c": 0, "nagi": 0, "ambros": 0, "oruni": 0, "david": 0, "weetman": 0, "martin": 0, "j": 0, "donnelli": 0, "2022": 0, "exploit": 0, "snakemak": [0, 2, 8, 12], "reveal": 0, "pattern": [0, 2], "insecticid": 0, "resist": 0, "malaria": 0, "vector": 0, "anophel": [0, 2], "gambia": [0, 2, 7], "biorxiv": 0, "06": 0, "17": 0, "493894": 0, "doi": 0, "http": [0, 8], "org": 0, "10": [0, 2], "1101": 0, "input": 0, "file": [0, 2, 12], "run": [0, 2, 8, 14], "modul": [0, 5, 7, 12], "imag": [1, 5], "wouldnt": [1, 5], "open": [1, 5, 8], "markdown": [1, 5], "from": [1, 2, 5], "ipython": [1, 5], "displai": [1, 5], "import": [1, 5, 8], "filenam": [1, 2, 5], "figur": 1, "chromosome_aim_karyo2": 1, "png": 1, "rna": [2, 3, 7, 12], "seq": [2, 3, 7, 12], "pop": [2, 3, 7, 12], "like": [2, 3], "all": [2, 8, 12], "should": [2, 7, 8], "config": [2, 7, 12], "yaml": [2, 7, 12], "The": [2, 7, 8], "look": 2, "follow": [2, 12], "metadata": [2, 5, 7], "sampl": [2, 7], "tsv": [2, 7], "samplesheet": 2, "dataset": [2, 7, 12], "ag_bouak": 2, "name": [2, 7], "can": [2, 7, 12], "anyth": 2, "us": [2, 7, 8, 12], "some": [2, 7], "main": 2, "output": 2, "fastq": [2, 7], "locat": [2, 7], "either": [2, 7], "specifi": [2, 7], "column": [2, 7], "fq1": [2, 7], "fq2": [2, 7], "pair": [2, 7], "end": [2, 7], "read": [2, 7], "OR": 2, "auto": 2, "true": 2, "gzip": [2, 7], "store": [2, 7], "resourc": [2, 7], "directori": [2, 12], "If": [2, 3, 7, 8, 12, 14], "kisumu1": 2, "kisumu1_1": 2, "gz": [2, 7], "kisumu1_2": 2, "treatment": [2, 7], "contain": [2, 7, 8], "underscor": 2, "set": [2, 12], "abov": 2, "rule": [2, 12], "path": [2, 7, 12], "give": 2, "ar": [2, 7], "sampleid_1": 2, "fals": 2, "point": [2, 7], "each": [2, 7], "chromosom": [2, 7], "appropri": 2, "speci": [2, 7], "current": 2, "variant": [2, 7], "analysi": [2, 7], "part": 2, "requir": [2, 7, 8], "gene": [2, 7], "map": [2, 7], "physic": 2, "pleas": [2, 3, 7, 8, 14], "ensur": 2, "correspond": [2, 7], "exactli": 2, "refer": [2, 7], "fasta": 2, "gff": 2, "extra": 2, "unwant": 2, "unplac": 2, "contig": [2, 7], "ignor": 2, "need": 2, "remov": 2, "prefix": 2, "Such": 2, "agamp4_": 2, "aaegl5_": 2, "sed": 2, "command": [2, 12], "e": 2, "": [2, 5], "an": [2, 3, 7, 8, 14], "gambiae_clean": 2, "2l": [2, 7], "2r": [2, 7], "3l": [2, 7], "3r": [2, 7], "x": [2, 7], "pest_chromosomes_agamp4": 2, "fa": [2, 7], "pest_transcripts_agamp4": 2, "12": 2, "pest_basefeatures_agamp4": 2, "gff3": [2, 7], "annot": [2, 7], "snpeffdb": 2, "anopheles_gambia": 2, "snpeff": [2, 7], "databas": [2, 7], "genes2transcript": 2, "examplegene2transcriptmap": [2, 7], "transcript": [2, 7], "list": [2, 7], "wish": [2, 7, 8, 12], "must": [2, 7, 8], "match": [2, 7], "featur": [2, 3, 7], "contrast": 2, "do": [2, 3, 7, 8], "fst": 2, "separ": [2, 7], "_": 2, "control_cas": 2, "kisumu_tiassal": 2, "themselv": 2, "kisumu_gambiaecont": 2, "gambiaecont_gambiaepm": 2, "cutadapt": [2, 7], "activ": 2, "adaptor": 2, "qualitycontrol": 2, "qualiti": 2, "control": 2, "pairwis": 2, "valu": [2, 7], "format": [2, 5, 7], "trim": [2, 7], "differentialexpress": 2, "venn": 2, "padj_threshold": 2, "0": 2, "05": 2, "group": [2, 7], "geneheatmap": 2, "gene_list_tsv": 2, "example_heatmap_gene_list": 2, "progressivegen": 2, "Will": 2, "find": [2, 14], "consist": 2, "up": [2, 12], "downregul": 2, "across": 2, "two": [2, 7], "comparison": 2, "For": [2, 7, 8], "exampl": [2, 7], "which": [2, 7, 8, 12], "upregul": 2, "kisumu": 2, "v": 2, "gambiaecont": 2, "gambiaepm": 2, "fc_threshold": 2, "1": [2, 5, 7], "kisumu_gambiaecont_gambiaepm": 2, "gsea": 2, "fgsea": 2, "enrich": 2, "gaf": 2, "vectorbas": 2, "50_agambiaepest_go": 2, "keggspeciesid": 2, "aga": 2, "3": [2, 5, 7], "letter": 2, "kegg": 2, "identifi": 2, "coluzzii": [2, 7], "replacestr": 2, "agap_": 2, "pathwai": 2, "have": [2, 3, 7, 8], "string": 2, "append": 2, "make": [2, 3], "diagram": 2, "de": 2, "variantanalysi": 2, "caller": 2, "freebay": 2, "ploidi": 2, "level": 2, "gener": [2, 8, 14], "we": [2, 3, 7, 8], "pool": 2, "diploid": 2, "2": [2, 5, 7], "number": 2, "individu": 2, "chunk": 2, "9": 2, "split": 2, "when": [2, 7, 8], "parallelis": 2, "less": 2, "recommend": [2, 8], "pca": [2, 7], "genotyp": 2, "missing": 2, "summarystatist": 2, "estim": 2, "summari": 2, "statist": 2, "dxy": 2, "pi": 2, "select": 2, "calcul": 2, "pb": 2, "per": 2, "window": 2, "branch": 2, "three": 2, "condit": 2, "close": 2, "relat": 2, "outgroup": 2, "gambiaepm_gambiaecont_kisumu": 2, "windownam": 2, "small": [2, 7], "larg": [2, 7], "windows": 2, "1000": 2, "5000": 2, "windowstep": 2, "500": 2, "l": 2, "onli": 2, "ancestri": 2, "inform": [2, 7, 8], "marker": 2, "aim": 2, "includ": 2, "git": [2, 3, 8], "repo": [2, 3, 7], "5": 2, "proport": 2, "between": 2, "gambcolu": 2, "gamb_vs_colu": 2, "zarr": 2, "arab": 2, "gambcolu_vs_arab": 2, "compkaryo": [2, 8], "determin": 2, "karyotyp": 2, "invers": 2, "2la": 2, "2rb": 2, "miscellan": 2, "creat": [2, 3, 8], "allel": 2, "count": 2, "xlsx": 2, "report": 2, "specif": 2, "mutat": 2, "choic": 2, "variantsofinterest": 2, "examplemut": 2, "genefamiliesheatmap": [2, 5], "eggnog": 2, "anogam_long": 2, "pep_eggnog_diamond": 2, "emapp": 2, "go": 2, "pfam": 2, "pep_pfamscan": 2, "lie": 2, "under": 2, "ag1000g": 2, "sweep": 2, "applic": 2, "encourag": [3, 8], "feedback": 3, "how": 3, "mai": [3, 7, 8, 14], "improv": 3, "get": [3, 5], "touch": 3, "email": [3, 14], "rais": [3, 5, 14], "issu": [3, 14], "github": [3, 7, 8], "fork": [3, 8], "repositori": [3, 8], "pull": 3, "request": 3, "addit": 3, "would": 3, "implemen": 3, "clone": [3, 8], "com": 3, "sanjaynagi": 3, "chang": 3, "commit": 3, "push": 3, "them": [3, 7], "back": 3, "local": 3, "result": [5, 7], "genediff": 5, "pdf": 5, "valueerror": 5, "traceback": 5, "most": 5, "recent": 5, "last": 5, "tmp": 5, "ipykernel_12544": 5, "3633508491": 5, "py": 5, "miniconda3": 5, "lib": 5, "python3": 5, "7": 5, "site": 5, "packag": [5, 8, 12], "core": [5, 12], "__init__": 5, "self": 5, "url": 5, "emb": 5, "width": 5, "height": 5, "retina": 5, "unconfin": 5, "1221": 5, "1222": 5, "_acceptable_embed": 5, "1223": 5, "cannot": 5, "1224": 5, "1225": 5, "_mimetyp": 5, "user": [7, 8], "provid": [7, 12], "tab": 7, "place": 7, "folder": 7, "default": 7, "examplesampl": 7, "shown": 7, "below": 7, "sampleid": 7, "strain": 7, "conttia1": 7, "conttia": 7, "tiassal": 7, "conttia2": 7, "conttia4": 7, "malatia1": 7, "malatia": 7, "malatia2": 7, "malatia4": 7, "In": [7, 8], "our": 7, "compar": 7, "relev": 7, "studi": 7, "same": 7, "defin": 7, "smaller": 7, "within": 7, "princip": 7, "compon": 7, "multipl": 7, "forward": 7, "revers": 7, "alreadi": 7, "wai": 7, "_1": 7, "_2": 7, "add": 7, "mean": 7, "anywher": 7, "access": 7, "arbitrari": 7, "dna": 7, "four": 7, "geneid": 7, "transcriptid": 7, "genenam": 7, "genedescript": 7, "necessari": [7, 8, 12], "connect": 7, "parent": 7, "ad": 7, "thei": 7, "These": 7, "entri": 7, "though": 7, "aed": 7, "aegypti": 7, "its": [7, 14], "simpli": 7, "explicitli": 7, "becaus": 7, "few": 7, "hundr": 7, "necessarili": 7, "want": [7, 8], "doe": 8, "howev": 8, "conda": [8, 12], "order": 8, "workflow": [8, 14], "sourc": 8, "manag": 8, "system": 8, "allow": 8, "varieti": 8, "program": 8, "whilst": 8, "softwar": 8, "environ": 8, "miniconda": 8, "io": 8, "project": 8, "en": 8, "latest": 8, "guid": 8, "index": 8, "html": 8, "reproduc": 8, "scalabl": 8, "written": 8, "itself": 8, "python": 8, "base": 8, "author": 8, "through": 8, "readthedoc": 8, "stabl": 8, "getting_start": 8, "next": 8, "step": 8, "think": 8, "contribut": 8, "account": 8, "prior": 8, "more": 8, "doc": 8, "submodul": 8, "mpileup2readcount": 8, "recurs": 8, "direct": 8, "automat": 8, "isol": 8, "so": 8, "other": 8, "onc": 12, "dry": 12, "see": 12, "n": 12, "indic": 12, "4": 12, "To": 12, "overrid": 12, "line": 12, "configfil": 12, "myconfig": 12, "test": 12, "problem": 14, "bug": 14, "tracker": 14, "guidanc": 14, "me": 14, "directli": 14, "consid": 14, "join": 14, "discord": 14, "server": 14}, "objects": {}, "objtypes": {}, "objnames": {}, "titleterms": {"rna": [0, 8, 11], "seq": [0, 8, 11], "pop": [0, 8, 11], "get": 0, "start": 0, "The": 0, "workflow": [0, 2, 3, 12], "troubleshoot": [0, 14], "contribut": [0, 3], "ancestri": 1, "configur": 2, "document": 2, "setup": 3, "differenti": 4, "express": 4, "isoform": 4, "level": 4, "gene": 5, "famili": 5, "heatmap": 5, "genet": 6, "divers": 6, "input": 7, "file": 7, "instal": 8, "depend": 8, "karyotyp": 9, "qualiti": 10, "control": 10, "modul": 11, "run": 12, "select": 13, "variant": 15, "call": 15}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.intersphinx": 1, "sphinxcontrib.bibtex": 9, "sphinx": 57}, "alltitles": {"RNA-Seq-Pop": [[0, "rna-seq-pop"], [8, "rna-seq-pop"]], "Getting started": [[0, null]], "The RNA-Seq-Pop workflow": [[0, null]], "Troubleshooting and contributing": [[0, null]], "Ancestry": [[1, "ancestry"]], "Configuring the workflow": [[2, "configuring-the-workflow"]], "documentation": [[2, "documentation"]], "Contributing to the workflow": [[3, "contributing-to-the-workflow"]], "Setup": [[3, "setup"]], "Differential expression": [[4, "differential-expression"]], "Isoform-level differential expression": [[4, "isoform-level-differential-expression"]], "Gene family heatmaps": [[5, "gene-family-heatmaps"]], "Genetic diversity": [[6, "genetic-diversity"]], "Input files": [[7, "input-files"]], "Installation": [[8, "installation"]], "Installing dependencies": [[8, "installing-dependencies"]], "Karyotyping": [[9, "karyotyping"]], "Quality control": [[10, "quality-control"]], "RNA-Seq-Pop modules": [[11, "rna-seq-pop-modules"]], "Running the workflow": [[12, "running-the-workflow"]], "Selection": [[13, "selection"]], "Troubleshooting": [[14, "troubleshooting"]], "Variant calling": [[15, "variant-calling"]]}, "indexentries": {}})