{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "703231a2",
   "metadata": {},
   "source": [
    "# Input files\n",
    "\n",
    "**Sample metadata file**\n",
    "\n",
    "The user must provide a tab-separated sample metadata file, which should be placed in the config/ folder and pointed to in the users `config.yaml`. The default name is `samples.tsv`. An example metadata file is provided in at `config/examplesamples.tsv`, and shown below: \n",
    "\n",
    "```\n",
    "samples.tsv:\n",
    "\n",
    "|sampleID|treatment|species  |strain  |\n",
    "|--------|---------|---------|--------|\n",
    "|ContTia1|ContTia  |coluzzii |Tiassale|\n",
    "|ContTia2|ContTia  |coluzzii |Tiassale|\n",
    "|ContTia4|ContTia  |coluzzii |Tiassale|\n",
    "|MalaTia1|MalaTia  |coluzzii |Tiassale|\n",
    "|MalaTia2|MalaTia  |coluzzii |Tiassale|\n",
    "|MalaTia4|MalaTia  |coluzzii |Tiassale|\n",
    "\n",
    "```\n",
    "\n",
    "In the `config.yaml`, we will use the treatment column to specify our comparative groups for analysis.\n",
    "\n",
    "\n",
    "If the strain information is not relevant to your study organism, please use the same values as for species. The strain column is used to define smaller groups within the data for principal components analysis (PCA), and is useful when analysing datasets with multiple strains.\n",
    "\n",
    "--- \n",
    "\n",
    "\n",
    "**Paired-end RNA-Sequencing fastq reads**\n",
    "\n",
    "Two gzipped fastq files for each sample are required, corresponding to the forward and reverse read. Reads can be already trimmed or RNA-Seq-Pop can trim them, using the cutadapt module.\n",
    "\n",
    "The read location may be specified in two ways:  \n",
    "\n",
    "1. Reads can be named as `{sampleID}_1.fastq.gz , {sampleID}_2.fastq.gz` and stored in `resources/reads/`. In the config.yaml, `fastq['auto'] == True`.\n",
    "\n",
    "   \n",
    "2. The user can add \"fq1\" and \"fq2\" columns to the `samples.tsv` metadata file, containing the path to each fastq file from the root rna-seq-pop directory. This means the fastq files can be stored anywhere accessible and have arbitrary naming. In the config.yaml, `fastq['auto'] == False`.\n",
    "\n",
    "---\n",
    "\n",
    "**Reference genome files**\n",
    "\n",
    "The user provides the path to the reference files in the configuration file (`config.yaml`).\n",
    "\n",
    "1. **Genome chromosomes reference file (.fa)**. Contains the DNA sequence for the genome.\n",
    "2. **Transcriptome reference file (.fa)**. Contains the DNA sequence for each transcript.\n",
    "3. **Genome feature file (.gff3 format)**. \n",
    "4. **Genes to Transcript mapping file (.tsv)**. An example is provided in the github repo (`resources/exampleGene2TranscriptMap.tsv`). This should contain four columns, GeneID, TranscriptID, GeneName, and GeneDescription, and is necessary for connecting transcripts to their parent genes, as well as adding gene annotations to results.\n",
    "   \n",
    "5. **SnpEff database name** (if performing variant calling).\n",
    "\n",
    "---\n",
    "\n",
    "**Contigs**\n",
    "\n",
    "The user must provide a list of contigs that they wish to analyse in the configuration file (`config.yaml`). These must match entries in the reference files (.fa, gff3). \n",
    "\n",
    "For example, in *An. gambiae*, we refer to the chromosomes as [\"2L\", \"2R\", \"3L\", \"3R\", \"X\"], though in Aedes aegypti, its simply [\"1\", \"2\", \"3\"].\n",
    "\n",
    "\n",
    "We specify contigs explicitly because in some organisms we have a few large chromosomes and then hundreds of small contigs, which we do not necessarily want to analyse."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8e5ee32",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
