
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Configuring the workflow</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script kind="utterances">

    var commentsRunWhenDOMLoaded = cb => {
    if (document.readyState != 'loading') {
        cb()
    } else if (document.addEventListener) {
        document.addEventListener('DOMContentLoaded', cb)
    } else {
        document.attachEvent('onreadystatechange', function() {
        if (document.readyState == 'complete') cb()
        })
    }
}

var addUtterances = () => {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src = "https://utteranc.es/client.js";
    script.async = "async";

    script.setAttribute("repo", "sanjaynagi/rna-seq-pop");
    script.setAttribute("issue-term", "pathname");
    script.setAttribute("theme", "github-light");
    script.setAttribute("label", "comment");
    script.setAttribute("crossorigin", "anonymous");

    sections = document.querySelectorAll("div.section");
    if (sections !== null) {
        section = sections[sections.length-1];
        section.appendChild(script);
    }
}
commentsRunWhenDOMLoaded(addUtterances);
</script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Running the workflow" href="running_the_workflow.html" />
    <link rel="prev" title="Input files" href="input_data.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../landing-page.html">
                    RNA-Seq-Pop
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="installation.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="input_data.html">
   Input files
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Configuring the workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="running_the_workflow.html">
   Running the workflow
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  The RNA-Seq-Pop workflow
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="rna-seq-pop.html">
   <strong>
    RNA-Seq-Pop tutorial
   </strong>
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="snakemake.html">
     Snakemake
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="quality_control.html">
     Quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="read_counts.html">
     Read counts
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="differential_expression.html">
     Differential expression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="gene_families.html">
     Gene family heatmaps
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="gsea.html">
     Gene set enrichment analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="variant_calling.html">
     Variant calling
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="genetic_diversity.html">
     Genetic diversity
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="variants_of_interest.html">
     Variants of Interest
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="selection.html">
     Selection
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="ancestry_informative_markers.html">
     Ancestry
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="karyotyping.html">
     Karyotyping
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Troubleshooting and contributing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="troubleshooting.html">
   Troubleshooting
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="contributing.html">
   Contributing to the workflow
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/sanjaynagi/rna-seq-pop"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/sanjaynagi/rna-seq-pop/issues/new?title=Issue%20on%20page%20%2Fnotebooks/configuring_the_workflow.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/notebooks/configuring_the_workflow.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Configuring the workflow</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section id="configuring-the-workflow">
<h1>Configuring the workflow<a class="headerlink" href="#configuring-the-workflow" title="Permalink to this heading">#</a></h1>
<p>Before running the RNA-Seq-Pop, we need to select which analyses we want to run (configuration). This is done by editing the file <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> in the <code class="docutils literal notranslate"><span class="pre">config</span></code> directory. The config file contains a number of options and parameters, which we describe below.</p>
<p>The configuration file looks like the following:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">metadata</span><span class="p">:</span> <span class="n">config</span><span class="o">/</span><span class="n">samples</span><span class="o">.</span><span class="n">tsv</span>                    

<span class="n">dataset</span><span class="p">:</span> <span class="s1">&#39;Ag_Bouake&#39;</span>                            

<span class="n">fastq</span><span class="p">:</span>
      <span class="n">auto</span><span class="p">:</span> <span class="kc">True</span>   
      <span class="n">paired</span><span class="p">:</span> <span class="kc">True</span>                                       
</pre></div>
</div>
<p><strong>metadata</strong><br />
Path to the sample metadata file, usually located in the <code class="docutils literal notranslate"><span class="pre">config</span></code> directory.<br />
<strong>dataset</strong><br />
Name for the dataset, which will be used to name some output files.<br />
<strong>fastq</strong><br />
<em>auto</em>: If true, fastq files are in resources/reads/ and following the <code class="docutils literal notranslate"><span class="pre">{sampleID_1.fastq.gz}</span></code> pattern. If False, there are fq1 and fq2 columns pointing to the reads files for each sample in the sample metadata file. If reads are stored in the <code class="docutils literal notranslate"><span class="pre">resources/reads</span></code> directory, the user must include this prefix in the fq1 and fq2 paths. The fq1 and fq2 columns are ignored if auto is True.<br />
<em>paired</em>: If True, the reads are paired-end. If False, the reads are single-end, and there is only an fq1 column in the sample metadata file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">contigs</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;2L&#39;</span><span class="p">,</span> <span class="s1">&#39;2R&#39;</span><span class="p">,</span> <span class="s1">&#39;3L&#39;</span><span class="p">,</span> <span class="s1">&#39;3R&#39;</span><span class="p">,</span> <span class="s1">&#39;X&#39;</span><span class="p">]</span>

<span class="n">reference</span><span class="p">:</span>
  <span class="n">genome</span><span class="p">:</span>
        <span class="s2">&quot;resources/reference/Anopheles-gambiae-PEST_CHROMOSOMES_AgamP4.fa&quot;</span>       
  <span class="n">transcriptome</span><span class="p">:</span>
        <span class="s2">&quot;resources/reference/Anopheles-gambiae-PEST_TRANSCRIPTS_AgamP4.12.fa&quot;</span>      
  <span class="n">gff</span><span class="p">:</span>
        <span class="s2">&quot;resources/reference/Anopheles-gambiae-PEST_BASEFEATURES_AgamP4.12.gff3&quot;</span>  
  <span class="n">snpeffdb</span><span class="p">:</span>
        <span class="s2">&quot;Anopheles_gambiae&quot;</span>                                                       
  <span class="n">genes2transcripts</span><span class="p">:</span>                                                                
        <span class="s2">&quot;resources/exampleGene2TranscriptMap.tsv&quot;</span>                                 
</pre></div>
</div>
<p><strong>contigs</strong><br />
A list of contigs you wish to analyse. Must match the reference genome files.</p>
<p><strong>Reference files</strong><br />
<em>genome</em>: Path to the genome (.fa)<br />
<em>transcriptome</em>: Path to the transcriptome (.fa)<br />
<em>gff</em>: Path to the genome feature file (.gff3)<br />
<em>genes2transcripts</em>: Path to the gene to transcript map (.tsv)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">contrasts</span><span class="p">:</span>
      <span class="o">-</span> <span class="s1">&#39;Kisumu_gambiaeCont&#39;</span>
      <span class="o">-</span> <span class="s1">&#39;gambiaeCont_gambiaePM&#39;</span>


<span class="n">QualityControl</span><span class="p">:</span>
      <span class="n">cutadapt</span><span class="p">:</span>
            <span class="n">activate</span><span class="p">:</span> <span class="kc">False</span>
            <span class="n">adaptors</span><span class="p">:</span>
      <span class="n">fastqc</span><span class="p">:</span>                               
            <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>
      <span class="n">mosdepth</span><span class="p">:</span>
            <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>
      <span class="n">multiqc</span><span class="p">:</span>
            <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>
</pre></div>
</div>
<p><strong>contrasts</strong><br />
A list of the pairwise contrasts you wish to run, from values in the treatment column of the sample metadata file. The format is <code class="docutils literal notranslate"><span class="pre">control_case</span></code>, or <code class="docutils literal notranslate"><span class="pre">susceptible_resistant</span></code>.</p>
<p><strong>QualityControl</strong><br />
<em>cutadapt</em>: Trim reads with cutadapt<br />
<em>adaptors</em>: adaptors to be trimmed</p>
<p><em>fastqc</em>: <em>activate</em>: Run fastqc on read data<br />
<em>mosdepth</em>: <em>activate</em>: Run mosdepth on read data<br />
<em>multiqc</em>: <em>activate</em>: Run multiqc on read data</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">DifferentialExpression</span><span class="p">:</span>
      <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>                          <span class="c1"># Activate differential expression analyses</span>

      <span class="n">venn</span><span class="p">:</span>
            <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>
            <span class="n">padj_threshold</span><span class="p">:</span> <span class="mf">0.05</span>
            <span class="n">groups</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;gambiaeCont_gambiaePM&#39;</span><span class="p">,</span> <span class="s1">&#39;Kisumu_gambiaeCont&#39;</span><span class="p">]</span>

      <span class="n">progressiveGenes</span><span class="p">:</span>                       
            <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>                    
            <span class="n">padj_threshold</span><span class="p">:</span> <span class="mf">0.05</span>
            <span class="n">fc_threshold</span><span class="p">:</span> <span class="mi">1</span>
            <span class="n">groups</span><span class="p">:</span> <span class="s2">&quot;Kisumu_gambiaeCont_gambiaePM&quot;</span>

      <span class="n">GSEA</span><span class="p">:</span>                                 <span class="c1"># Activate fgsea Gene set enrichment analysis </span>
            <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>
            <span class="n">gaf</span><span class="p">:</span> <span class="s2">&quot;resources/reference/VectorBase-50_AgambiaePEST_GO.gaf&quot;</span>
</pre></div>
</div>
<p><strong>Differential Expression</strong><br />
<em>activate</em>: Run differential expression analyses</p>
<p><strong>Venn</strong><br />
<em>activate</em>: Generate venn diagrams of DE expressed genes<br />
<em>padj_threshold</em>: threshold to call significance<br />
<em>groups</em>: string or list of strings to indicate the DE comparisons we wish to make a venn for.</p>
<p><strong>ProgressiveGenes</strong><br />
<em>activate</em>: Run progressiveGenes analysis. Will find genes that are consistently up or downregulated across two comparisons.<br />
<em>padj_threshold</em>: threshold to call significance<br />
<em>fc_threshold</em>: threshold to call overexpression<br />
<em>groups</em>: String or list of strings, which indicate the 3 populations and 2 DE comparisons to be compared. In this example - which genes upregulated in Kisumu v gambiaeCont, are also upregulated in gambiaeCont v gambiaePM.</p>
<p><strong>gsea</strong><br />
<em>activate</em>: run hypergeometric test enrichment analysis.<br />
<em>gaf</em>: gaf file with go annotations.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">VariantAnalysis</span><span class="p">:</span>
      <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>
      <span class="n">ploidy</span><span class="p">:</span> <span class="mi">10</span>                          
      <span class="n">chunks</span><span class="p">:</span> <span class="mi">9</span>                           

      <span class="n">pca</span><span class="p">:</span>                                <span class="c1"># Run PCA on the genotype data?</span>
            <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>
            <span class="n">missingness</span><span class="p">:</span> <span class="mi">1</span> 
      
      <span class="n">summaryStatistics</span><span class="p">:</span>                  <span class="c1"># Estimate Population Genetic Summary Statistics such as Dxy, Pi</span>
            <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>
            <span class="n">missingness</span><span class="p">:</span> <span class="mi">1</span>
      
      <span class="n">selection</span><span class="p">:</span>                          <span class="c1"># Calculate Fst and PBS per gene and in windows</span>
            <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>
            <span class="n">missingness</span><span class="p">:</span> <span class="mi">1</span>

            <span class="n">pbs</span><span class="p">:</span>
                  <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>          
                  <span class="n">contrasts</span><span class="p">:</span> 
                        <span class="o">-</span> <span class="s1">&#39;gambiaePM_gambiaeCont_Kisumu&#39;</span>
</pre></div>
</div>
<p><strong>VariantAnalysis</strong><br />
<em>activate</em> : Perform genome alignment and variant calling.<br />
<em>ploidy</em> : the ploidy of our samples to call variants at. If you have pooled data, the ploidy should be n_samples per pool * organism ploidy.<br />
<em>chunks</em> : The number of chunks to split each chromosome into to parallelise variant calling.</p>
<p><strong>pca</strong><br />
<em>activate</em>: Run PCA on the SNP data<br />
<em>missingness</em>: a filter - the proportion of samples that have data at a given allele. 1 means a genomic position can have no missing calls across all samples.</p>
<p><strong>summaryStatistics</strong><br />
<em>activate</em>: Run genetic diversity analyses.<br />
<em>missingness</em>: a filter - the proportion of samples that have data at a given allele. 1 means a genomic position can have no missing calls across all samples.</p>
<p><strong>selection</strong><br />
<em>activate</em>: Run Fst analysis<br />
<em>missingness</em>: a filter - the proportion of samples that have data at a given allele. 1 means a genomic position can have no missing calls across all samples.</p>
<p><strong>pbs</strong><br />
<em>activate</em>: Run population branch statistic analysis. Requires 3 suitable treatment groups, two closely related and an outgroup. For resistance, do survivors_unexposed_susceptible.<br />
<em>contrasts</em>: List of the 3 groups as strings separated by an underscore.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>      <span class="n">ancestry</span><span class="p">:</span>                                
            <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>
            <span class="n">missingness</span><span class="p">:</span> <span class="mf">0.5</span>                                <span class="c1"># proportion between 0 and 1 </span>
            <span class="n">gambcolu</span><span class="p">:</span> <span class="s2">&quot;resources/gamb_vs_colu.zarr&quot;</span>         <span class="c1"># path to gambcolu AIMs</span>
            <span class="n">arab</span><span class="p">:</span> <span class="s2">&quot;resources/gambcolu_vs_arab.zarr&quot;</span>         <span class="c1"># path to arab AIMs</span>

      <span class="n">karyotype</span><span class="p">:</span>                            
            <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>
            <span class="n">inversions</span><span class="p">:</span>
                  <span class="o">-</span> <span class="s2">&quot;2La&quot;</span>
                  <span class="o">-</span> <span class="s2">&quot;2Rb&quot;</span>
</pre></div>
</div>
<p><strong>ancestry</strong> (<em>An. gambiae</em> only)<br />
<em>activate</em>: If True, run AIM analysis.<br />
<em>missingness</em>: a filter - the proportion of samples that have data at a given allele. 1 means a genomic position can have no missing calls across all samples.<br />
<em>gambcolu</em>: path to gamb_colu aims<br />
<em>arab</em>: path to gambcolu_vs_arabiensis aims</p>
<p><strong>karyotype</strong> (<em>An. gambiae</em> only)<br />
<em>activate</em>: If True, run karyotyping module.<br />
<em>inversions</em>: A list of inversions in compKaryo to analyse.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">miscellaneous</span><span class="p">:</span>
      <span class="n">VariantsOfInterest</span><span class="p">:</span>             
            <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>                             
            <span class="n">path</span><span class="p">:</span> <span class="s2">&quot;resources/exampleMutations.tsv&quot;</span>         
      
      <span class="n">GeneFamiliesHeatmap</span><span class="p">:</span>
            <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>
            <span class="n">eggnog</span><span class="p">:</span> <span class="n">resources</span><span class="o">/</span><span class="n">Anogam_long</span><span class="o">.</span><span class="n">pep_eggnog_diamond</span><span class="o">.</span><span class="n">emapper</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">GO</span>
            <span class="n">pfam</span><span class="p">:</span> <span class="n">resources</span><span class="o">/</span><span class="n">Anogam_long</span><span class="o">.</span><span class="n">pep_Pfamscan</span><span class="o">.</span><span class="n">seqs</span>

      <span class="n">sweeps</span><span class="p">:</span>
            <span class="n">activate</span><span class="p">:</span> <span class="kc">True</span>
            <span class="n">padj_threshold</span><span class="p">:</span> <span class="mf">0.05</span>
            <span class="n">fc_threshold</span><span class="p">:</span> <span class="mf">1.5</span>
</pre></div>
</div>
<p><strong>VariantsOfInterest</strong><br />
<em>activate</em>: If True, run variants of interest analysis<br />
<em>path</em>: Path to variants of interest data</p>
<p><strong>GeneFamiliesHeatmap</strong><br />
<em>activate</em>: Run heatmaps on gene families using go terms and pfam domains.<br />
<em>eggnog</em>: path to eggnog file with go annotations for your organism (produced by eggnog-mapper).<br />
<em>pfam</em>: path to pfam domains for your organism.</p>
<p><strong>sweeps</strong> (<em>An. gambiae</em> only)<br />
<em>activate</em>: Run analysis to determine if DE genes lie under known selective sweep loci.<br />
<em>padj_threshold</em>: threshold to call significance<br />
<em>fc_threshold</em>: threshold to call overexpression</p>
<p><strong>If you have any issues configuring the pipeline, please watch the video walkthrough first, and raise an issue on github or <a class="reference external" href="mailto:sanjay&#46;c&#46;nagi&#37;&#52;&#48;gmail&#46;com">email me</a>.</strong></p>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./notebooks"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="input_data.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Input files</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="running_the_workflow.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Running the workflow</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Sanjay Curtis Nagi<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>