

<!DOCTYPE html>


<html >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Input files</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script kind="utterances">

    var commentsRunWhenDOMLoaded = cb => {
    if (document.readyState != 'loading') {
        cb()
    } else if (document.addEventListener) {
        document.addEventListener('DOMContentLoaded', cb)
    } else {
        document.attachEvent('onreadystatechange', function() {
        if (document.readyState == 'complete') cb()
        })
    }
}

var addUtterances = () => {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src = "https://utteranc.es/client.js";
    script.async = "async";

    script.setAttribute("repo", "sanjaynagi/rna-seq-pop");
    script.setAttribute("issue-term", "pathname");
    script.setAttribute("theme", "github-light");
    script.setAttribute("label", "comment");
    script.setAttribute("crossorigin", "anonymous");

    sections = document.querySelectorAll("div.section");
    if (sections !== null) {
        section = sections[sections.length-1];
        section.appendChild(script);
    }
}
commentsRunWhenDOMLoaded(addUtterances);
</script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/input_data';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Configuring the workflow" href="configuring_the_workflow.html" />
    <link rel="prev" title="Installation" href="installation.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="None"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../landing-page.html">
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../landing-page.html">
                    RNA-Seq-Pop
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Input files</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuring_the_workflow.html">Configuring the workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="running_the_workflow.html">Running the workflow</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">The RNA-Seq-Pop workflow</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="rna-seq-pop.html"><strong>RNA-Seq-Pop tutorial</strong></a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="snakemake.html">Snakemake</a></li>
<li class="toctree-l2"><a class="reference internal" href="quality_control.html">Quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="read_counts.html">Counts</a></li>
<li class="toctree-l2"><a class="reference internal" href="differential_expression.html">Differential expression</a></li>
<li class="toctree-l2"><a class="reference internal" href="gene_families.html">Gene family heatmaps</a></li>
<li class="toctree-l2"><a class="reference internal" href="gsea.html">Gene set enrichment analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="variant_calling.html">Variant calling</a></li>
<li class="toctree-l2"><a class="reference internal" href="genetic_diversity.html">Genetic diversity</a></li>
<li class="toctree-l2"><a class="reference internal" href="variants_of_interest.html">Variants of Interest</a></li>
<li class="toctree-l2"><a class="reference internal" href="selection.html">Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="ancestry_informative_markers.html">Ancestry</a></li>
<li class="toctree-l2"><a class="reference internal" href="karyotyping.html">Karyotyping</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Troubleshooting and contributing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="troubleshooting.html">Troubleshooting</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing to the workflow</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/sanjaynagi/rna-seq-pop" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/sanjaynagi/rna-seq-pop/issues/new?title=Issue%20on%20page%20%2Fnotebooks/input_data.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/notebooks/input_data.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Input files</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="input-files">
<h1>Input files<a class="headerlink" href="#input-files" title="Permalink to this headline">#</a></h1>
<p>RNA-Seq-Pop requires only:</p>
<ul class="simple">
<li><p>Sample metadata file <code class="docutils literal notranslate"><span class="pre">.tsv</span></code></p></li>
<li><p>Single-end or Paired-end RNA-Seq data - <code class="docutils literal notranslate"><span class="pre">.fastq.gz</span></code></p></li>
<li><p>Reference genome files</p></li>
</ul>
<p><strong>Sample metadata file</strong></p>
<p>The user must provide a tab-separated sample metadata file, which should be placed in the config/ folder and pointed to in the users <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>. The default name is <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code>. An example metadata file is provided in at <code class="docutils literal notranslate"><span class="pre">config/examplesamples.tsv</span></code>, and shown below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samples</span><span class="o">.</span><span class="n">tsv</span><span class="p">:</span>

<span class="o">|</span><span class="n">sampleID</span><span class="o">|</span><span class="n">treatment</span><span class="o">|</span><span class="n">species</span>  <span class="o">|</span><span class="n">strain</span>  <span class="o">|</span>
<span class="o">|--------|---------|---------|--------|</span>
<span class="o">|</span><span class="n">ContTia1</span><span class="o">|</span><span class="n">ContTia</span>  <span class="o">|</span><span class="n">coluzzii</span> <span class="o">|</span><span class="n">Tiassale</span><span class="o">|</span>
<span class="o">|</span><span class="n">ContTia2</span><span class="o">|</span><span class="n">ContTia</span>  <span class="o">|</span><span class="n">coluzzii</span> <span class="o">|</span><span class="n">Tiassale</span><span class="o">|</span>
<span class="o">|</span><span class="n">ContTia4</span><span class="o">|</span><span class="n">ContTia</span>  <span class="o">|</span><span class="n">coluzzii</span> <span class="o">|</span><span class="n">Tiassale</span><span class="o">|</span>
<span class="o">|</span><span class="n">MalaTia1</span><span class="o">|</span><span class="n">MalaTia</span>  <span class="o">|</span><span class="n">coluzzii</span> <span class="o">|</span><span class="n">Tiassale</span><span class="o">|</span>
<span class="o">|</span><span class="n">MalaTia2</span><span class="o">|</span><span class="n">MalaTia</span>  <span class="o">|</span><span class="n">coluzzii</span> <span class="o">|</span><span class="n">Tiassale</span><span class="o">|</span>
<span class="o">|</span><span class="n">MalaTia4</span><span class="o">|</span><span class="n">MalaTia</span>  <span class="o">|</span><span class="n">coluzzii</span> <span class="o">|</span><span class="n">Tiassale</span><span class="o">|</span>

</pre></div>
</div>
<p>In the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>, we will use the treatment column to specify our comparative groups for analysis.</p>
<p>If the strain information is not relevant to your study organism, please use the same values as for species. The strain column is used to define smaller groups within the data for principal components analysis (PCA), and is useful when analysing datasets with multiple strains.</p>
<hr class="docutils" />
<p><strong>Single or Paired-end RNA-Sequencing fastq reads</strong></p>
<p>One or two gzipped fastq files for each sample are required, depending on whether the user is using single-end or paired-end fastq files. Reads can be already trimmed or RNA-Seq-Pop can trim them, using the cutadapt module.</p>
<p>The read location may be specified in two ways:</p>
<ol class="arabic simple">
<li><p>Reads can be named as <code class="docutils literal notranslate"><span class="pre">{sampleID}_1.fastq.gz</span></code>, <code class="docutils literal notranslate"><span class="pre">{sampleID}_2.fastq.gz</span></code> and stored in <code class="docutils literal notranslate"><span class="pre">resources/reads/</span></code>. In the config.yaml, <code class="docutils literal notranslate"><span class="pre">fastq['auto']</span> <span class="pre">==</span> <span class="pre">True</span></code>, meaning snakemake will look for files in this folder which follow this naming pattern. For single-end reads, only the first <code class="docutils literal notranslate"><span class="pre">_1.fastq.gz</span></code> file is required.</p></li>
<li><p>The user can add “fq1” and “fq2” columns to the <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> metadata file, containing the path to each fastq file from the root rna-seq-pop directory. This allows the fastq files to be stored anywhere that is accessible and have arbitrary naming. In the config.yaml, this option is  <code class="docutils literal notranslate"><span class="pre">fastq['auto']</span> <span class="pre">==</span> <span class="pre">False</span></code>. For single-end reads, only the “fq1” column is required.</p></li>
</ol>
<hr class="docutils" />
<p><strong>Reference genome files</strong></p>
<p>RNA-Seq-Pop uses kallisto to perform differential expression analysis, which takes as input a reference transcriptome. The user must provide a fasta file containing the transcriptome sequence in fasta format. For variant calling, genome alignment is performed with hisat2, which requires a fasta file containing the genome sequence. All input <code class="docutils literal notranslate"><span class="pre">.fa</span></code> files can be gzipped <code class="docutils literal notranslate"><span class="pre">.fa.gz</span></code>.</p>
<p>The user provides the path to the reference files in the configuration file (<code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>).</p>
<ol class="arabic simple">
<li><p><strong>Genome chromosomes reference file (.fa/.fa.gz)</strong>. Contains the DNA sequence for the genome in fasta format.</p></li>
<li><p><strong>Transcriptome reference file (.fa/fa.gz)</strong>. Contains the DNA sequence for each transcript in fasta format.</p></li>
<li><p><strong>Genome feature file (.gff3 format)</strong>.</p></li>
<li><p><strong>Genes to Transcript mapping file (.tsv)</strong>. An example is provided in the github repo (<code class="docutils literal notranslate"><span class="pre">resources/exampleGene2TranscriptMap.tsv</span></code>). This should contain four columns, GeneID, TranscriptID, GeneName, and GeneDescription, and is necessary for connecting transcripts to their parent genes, as well as adding gene annotations to results. Files for <em>Anopheles gambiae, funestus</em> and <em>Aedes aegypti</em> are provided in the github repo.</p></li>
<li><p><strong>SnpEff database name</strong> (if performing variant calling).</p></li>
</ol>
<hr class="docutils" />
<p><strong>Contigs</strong></p>
<p>The user must provide a list of contigs that they wish to analyse in the configuration file (<code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>). These must match entries in the reference files (.fa, gff3).</p>
<p>For example, in <em>An. gambiae</em>, we refer to the chromosomes as [“2L”, “2R”, “3L”, “3R”, “X”], though in Aedes aegypti, its simply [“1”, “2”, “3”].</p>
<p>We specify contigs explicitly because in some reference assemblies we have a few large full chromosomes, plus hundreds of small contigs, which we may not necessarily wish to analyse.</p>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./notebooks"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="installation.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Installation</p>
      </div>
    </a>
    <a class="right-next"
       href="configuring_the_workflow.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Configuring the workflow</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Sanjay Curtis Nagi
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>