
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Input files</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script kind="utterances">

    var commentsRunWhenDOMLoaded = cb => {
    if (document.readyState != 'loading') {
        cb()
    } else if (document.addEventListener) {
        document.addEventListener('DOMContentLoaded', cb)
    } else {
        document.attachEvent('onreadystatechange', function() {
        if (document.readyState == 'complete') cb()
        })
    }
}

var addUtterances = () => {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src = "https://utteranc.es/client.js";
    script.async = "async";

    script.setAttribute("repo", "sanjaynagi/rna-seq-pop");
    script.setAttribute("issue-term", "pathname");
    script.setAttribute("theme", "github-light");
    script.setAttribute("label", "comment");
    script.setAttribute("crossorigin", "anonymous");

    sections = document.querySelectorAll("div.section");
    if (sections !== null) {
        section = sections[sections.length-1];
        section.appendChild(script);
    }
}
commentsRunWhenDOMLoaded(addUtterances);
</script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Configuring the workflow" href="configuring_the_workflow.html" />
    <link rel="prev" title="Installation" href="installation.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../landing-page.html">
                    RNA-Seq-Pop
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="installation.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Input files
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="configuring_the_workflow.html">
   Configuring the workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="running_the_workflow.html">
   Running the workflow
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  The RNA-Seq-Pop workflow
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="rna-seq-pop.html">
   <strong>
    RNA-Seq-Pop tutorial
   </strong>
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="snakemake.html">
     Snakemake
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="quality_control.html">
     Quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="read_counts.html">
     Read counts
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="differential_expression.html">
     Differential expression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="gene_families.html">
     Gene family heatmaps
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="gsea.html">
     Gene set enrichment analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="variant_calling.html">
     Variant calling
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="genetic_diversity.html">
     Genetic diversity
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="variants_of_interest.html">
     Variants of Interest
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="selection.html">
     Selection
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="ancestry_informative_markers.html">
     Ancestry
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="karyotyping.html">
     Karyotyping
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Troubleshooting and contributing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="troubleshooting.html">
   Troubleshooting
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="contributing.html">
   Contributing to the workflow
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/sanjaynagi/rna-seq-pop"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/sanjaynagi/rna-seq-pop/issues/new?title=Issue%20on%20page%20%2Fnotebooks/input_data.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/notebooks/input_data.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Input files</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="input-files">
<h1>Input files<a class="headerlink" href="#input-files" title="Permalink to this heading">#</a></h1>
<p>RNA-Seq-Pop requires only:</p>
<ul class="simple">
<li><p>Sample metadata file <code class="docutils literal notranslate"><span class="pre">.tsv</span></code></p></li>
<li><p>Single-end or Paired-end RNA-Seq data - <code class="docutils literal notranslate"><span class="pre">.fastq.gz</span></code></p></li>
<li><p>Reference genome files</p></li>
</ul>
<p><strong>Sample metadata file</strong></p>
<p>The user must provide a tab-separated sample metadata file, which should be placed in the config/ folder and pointed to in the users <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>. The default name is <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code>. An example metadata file is provided in at <code class="docutils literal notranslate"><span class="pre">config/examplesamples.tsv</span></code>, and shown below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samples</span><span class="o">.</span><span class="n">tsv</span><span class="p">:</span>

<span class="o">|</span><span class="n">sampleID</span><span class="o">|</span><span class="n">treatment</span><span class="o">|</span><span class="n">species</span>  <span class="o">|</span><span class="n">strain</span>  <span class="o">|</span>
<span class="o">|--------|---------|---------|--------|</span>
<span class="o">|</span><span class="n">ContTia1</span><span class="o">|</span><span class="n">ContTia</span>  <span class="o">|</span><span class="n">coluzzii</span> <span class="o">|</span><span class="n">Tiassale</span><span class="o">|</span>
<span class="o">|</span><span class="n">ContTia2</span><span class="o">|</span><span class="n">ContTia</span>  <span class="o">|</span><span class="n">coluzzii</span> <span class="o">|</span><span class="n">Tiassale</span><span class="o">|</span>
<span class="o">|</span><span class="n">ContTia4</span><span class="o">|</span><span class="n">ContTia</span>  <span class="o">|</span><span class="n">coluzzii</span> <span class="o">|</span><span class="n">Tiassale</span><span class="o">|</span>
<span class="o">|</span><span class="n">MalaTia1</span><span class="o">|</span><span class="n">MalaTia</span>  <span class="o">|</span><span class="n">coluzzii</span> <span class="o">|</span><span class="n">Tiassale</span><span class="o">|</span>
<span class="o">|</span><span class="n">MalaTia2</span><span class="o">|</span><span class="n">MalaTia</span>  <span class="o">|</span><span class="n">coluzzii</span> <span class="o">|</span><span class="n">Tiassale</span><span class="o">|</span>
<span class="o">|</span><span class="n">MalaTia4</span><span class="o">|</span><span class="n">MalaTia</span>  <span class="o">|</span><span class="n">coluzzii</span> <span class="o">|</span><span class="n">Tiassale</span><span class="o">|</span>

</pre></div>
</div>
<p>In the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>, we will use the treatment column to specify our comparative groups for analysis.</p>
<p>If the strain information is not relevant to your study organism, please use the same values as for species. The strain column is used to define smaller groups within the data for principal components analysis (PCA), and is useful when analysing datasets with multiple strains.</p>
<hr class="docutils" />
<p><strong>Single or Paired-end RNA-Sequencing fastq reads</strong></p>
<p>One or two gzipped fastq files for each sample are required, depending on whether the user is using single-end or paired-end fastq files. Reads can be already trimmed or RNA-Seq-Pop can trim them, using the cutadapt module.</p>
<p>The read location may be specified in two ways:</p>
<ol class="simple">
<li><p>Reads can be named as <code class="docutils literal notranslate"><span class="pre">{sampleID}_1.fastq.gz</span></code>, <code class="docutils literal notranslate"><span class="pre">{sampleID}_2.fastq.gz</span></code> and stored in <code class="docutils literal notranslate"><span class="pre">resources/reads/</span></code>. In the config.yaml, <code class="docutils literal notranslate"><span class="pre">fastq['auto']</span> <span class="pre">==</span> <span class="pre">True</span></code>, meaning snakemake will look for files in this folder which follow this naming pattern. For single-end reads, only the first <code class="docutils literal notranslate"><span class="pre">_1.fastq.gz</span></code> file is required.</p></li>
<li><p>The user can add “fq1” and “fq2” columns to the <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> metadata file, containing the path to each fastq file from the root rna-seq-pop directory. This allows the fastq files to be stored anywhere that is accessible and have arbitrary naming. In the config.yaml, this option is  <code class="docutils literal notranslate"><span class="pre">fastq['auto']</span> <span class="pre">==</span> <span class="pre">False</span></code>. For single-end reads, only the “fq1” column is required.</p></li>
</ol>
<hr class="docutils" />
<p><strong>Reference genome files</strong></p>
<p>RNA-Seq-Pop uses kallisto to perform differential expression analysis, which takes as input a reference transcriptome. The user must provide a fasta file containing the transcriptome sequence in fasta format. For variant calling, genome alignment is performed with hisat2, which requires a fasta file containing the genome sequence. All input <code class="docutils literal notranslate"><span class="pre">.fa</span></code> files can be gzipped <code class="docutils literal notranslate"><span class="pre">.fa.gz</span></code>.</p>
<p>The user provides the path to the reference files in the configuration file (<code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>).</p>
<ol class="simple">
<li><p><strong>Genome chromosomes reference file (.fa/.fa.gz)</strong>. Contains the DNA sequence for the genome in fasta format.</p></li>
<li><p><strong>Transcriptome reference file (.fa/fa.gz)</strong>. Contains the DNA sequence for each transcript in fasta format.</p></li>
<li><p><strong>Genome feature file (.gff3 format)</strong>.</p></li>
<li><p><strong>Genes to Transcript mapping file (.tsv)</strong>. An example is provided in the github repo (<code class="docutils literal notranslate"><span class="pre">resources/exampleGene2TranscriptMap.tsv</span></code>). This should contain four columns, GeneID, TranscriptID, GeneName, and GeneDescription, and is necessary for connecting transcripts to their parent genes, as well as adding gene annotations to results. Files for <em>Anopheles gambiae, funestus</em> and <em>Aedes aegypti</em> are provided in the github repo.</p></li>
<li><p><strong>SnpEff database name</strong> (if performing variant calling).</p></li>
</ol>
<hr class="docutils" />
<p><strong>Contigs</strong></p>
<p>The user must provide a list of contigs that they wish to analyse in the configuration file (<code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>). These must match entries in the reference files (.fa, gff3).</p>
<p>For example, in <em>An. gambiae</em>, we refer to the chromosomes as [“2L”, “2R”, “3L”, “3R”, “X”], though in Aedes aegypti, its simply [“1”, “2”, “3”].</p>
<p>We specify contigs explicitly because in some reference assemblies we have a few large full chromosomes, plus hundreds of small contigs, which we may not necessarily wish to analyse.</p>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./notebooks"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="installation.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Installation</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="configuring_the_workflow.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Configuring the workflow</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Sanjay Curtis Nagi<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>